### PR TITLE
implement a simple mut-to-const pass

### LINKED ISSUE #69: OsrMut should take an expr, not a var
https://github.com/reactorlabs/sourir/blob/master/instr.ml#L49

We want:
```ocaml
and osr_def =
  | OsrConst of variable * expression
  | OsrMut of variable * expression (* This was variable *)
  | OsrMutUndef of variable
```

There was probably a reason why it wasn't like this, but we should investigate.

### PR DESCRIPTION
**Pull request by o-**
to reduce the friction I changed osr to take an expression in the mut case. This resolves #69 en passant.

### CODE REVIEW

**Comment by gasche** [comment]
I still dislike the idea that osr would always have to create a fresh mutable variable, instead of being able to alias an existing heap location. I think the change is a regression here -- see #45 and #48. We can have just `mut x = y` or have both `mut x <- y` and `mut x = e`, but having only the latter is meh.

**Comment by o-** [comment]
good point. will do...

**Comment by o-** [comment]
actually I don't think it's a good idea to have both `mut x = 1` and `mut x <- 1`.

Either we think osr should not change the heap in general (then we must have some other solution to the problem).

Or we accept that it does. In this case having the two osr cases just complicates matters for no good reason. I think whether the osr reuses a heap reference or not is an implementation detail.

In the function call case the distinction is relevant since one creates an alias and the other doesn't. But in this case the old reference dies. Therefore the two are equivalent. It's just an optimization that the osr mechanism can reuse the reference in one of the cases.

**Comment by o-** [comment]
retraction. Writing the above comment made me realize that osr-by-value of an aliased mutable variable breaks the semantics! So we need the two cases

**Comment by o-** [comment]
ptal

File: transform_constantfold.ml [hunk 1]
```diff
@@ -104,3 +106,55 @@ let const_prop ({formals; instrs} : analysis_input) : instructions option =
...
+
+let make_constant (({formals; instrs} as inp) : analysis_input) : instructions option =
+  let required = Analysis.required inp in
+  let constant var pc =
+    match[@warning "-4"] instrs.(pc) with
+    | Assign (x, _) when x = var -> false
+    | Drop x when x = var -> false
+    | Clear x when x = var -> false
+    | Read x when x = var -> false
+    | Call (_, _,exp) ->
```
Comment (mhyee) [inline transform_constantfold.ml]: Nitpick: missing space before `exp`.

**Comment by mhyee** [comment]
Looks good to me.

**Comment by gasche** [comment]
I'm sorry for the nitpick, but a minor thing is itching me with the merged version: I would rather expect `mut x <- y` and `mut x = e`, because the latter looks like the declaration syntax (so it makes sense for it to be general (accept expressions) and have the same semantics), while the former looks more like "x is replaced by y" to me, which (to me) suggests the reuse of the heap cell. Does someone have a reason of taste for the current choice which, I believe, is the opposite one, `mut x <- e` and `mut x = y`?

**Comment by o-** [comment]
hmm I don't really have a good reason for picking one over the other now that I think about it. Honestly I think there lies no good answer here. The existing `<-` does not declare a variable, but updates it (also `a <- b` does not create an alias but assignes "by-value"). Which does not correspond to either of the osr cases. To be honest the most consistent with what we have would be `mut x = e` and `mut x = &y`. If you don't like the `&` we should find a new syntax for both the args and the osr, but I feel like they should be consistent.

**Comment by o-** [comment]
let's continue at #92

**Pull request merged**